docs/architecture/adr-069-unordered-account.md

```diff
@@ -0,0 +1,177 @@
+
+  _, ok = dtm.hashes[hash]
+  return
+}
+
+func (dtm *DedupTxHashManager) Size() int {
+  dtm.mutex.RLock()
+  defer dtm.mutex.RUnlock()
+
+  return len(dtm).hashes
+}
+
+func (dtm *DedupTxHashManager) Add(hash TxHash, expire uint64) (ok bool) {
+  dtm.mutex.Lock()
+  defer dtm.mutex.Unlock()
+
+  dtm.hashes[hash] = expire
+  return
+}
+
+// EndBlock remove expired tx hashes, need to wire in abci cycles.
+func (dtm *DedupTxHashManager) EndBlock(ctx sdk.Context) {
+  if ctx.BlockNumber() % ExpireCheckInterval != 0 {
+    return
+  }
+
+  dtm.mutex.Lock()
+  defer dtm.mutex.Unlock()
+
+  for k, expire := range dtm.hashes {
+    if ctx.BlockNumber() > expire {
+      delete(dtm.hashes, k)
+    }
+  }
+}
+```
+
+### Ante Handlers
+
+Bypass the nonce decorator for un-ordered transactions.
+
+```golang
+func (isd IncrementSequenceDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  if tx.UnOrdered() {
+    return next(ctx, tx, simulate)
+  }
+  
+  // the previous logic
+}
+```
+
+A decorator for the new logic.
+
+```golang
+type TxHash [32]byte
+
+const (
+  // MaxNumberOfTxHash * 32 = 128M max memory usage
+  MaxNumberOfTxHash = 1024 * 1024 * 4
+
+  // MaxUnOrderedTTL defines the maximum ttl an un-order tx can set
+  MaxUnOrderedTTL = 1024
+)
+
+type DedupTxDecorator struct {
+  m *DedupTxHashManager
+}
+
+func (dtd *DedupTxDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  // only apply to un-ordered transactions
+  if !tx.UnOrdered() {
+    return next(ctx, tx, simulate)
+  }
+
+  if tx.TimeoutHeight() == 0 {
+    return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "unordered tx must set timeout-height")
+  }
+
+  if tx.TimeoutHeight() > ctx.BlockHeight() + MaxUnOrderedTTL {
+    return nil, errorsmod.Wrapf(sdkerrors.ErrLogic, "unordered tx ttl exceeds %d", MaxUnOrderedTTL)
+  }
+
+  if !ctx.IsCheckTx() {
+    // a new tx included in the block, add the hash to the dictionary
+    if dtd.m.Size() >= MaxNumberOfTxHash {
+      return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "dedup map is full")
+    }
+    dtd.m.Add(tx.Hash(), tx.TimeoutHeight())
+  } else {
+    // check for duplicates
+    if dtd.m.Contains(tx.Hash()) {
+      return nil, errorsmod.Wrap(sdkerrors.ErrLogic, "tx is duplicated")
+    }
+  }
+
+  return next(ctx, tx, simulate)
+}
+```
+
+### EndBlocker
+
+Wire up the `EndBlock` method of `DedupTxHashManager` into the application's abci life cycle.
+
+### Start Up
+
+On start up, the node needs to re-fill the tx hash dictionary of `DedupTxHashManager` by scanning `MaxUnOrderedTTL` number of historical blocks for un-ordered transactions.
+
+An alternative design is to store the tx hash dictionary in kv store, then no need to warm up on start up.
+
+## Consequences
+
+### Positive
+
+* Support un-ordered and concurrent transaction inclusion.
+
+### Negative
+
+- Start up overhead to scan historical blocks.
```

Comment:
should we use the filesystem to write the known hashes at shutdown? On start we would populate from the file

Reply:
As long as we write to disk, we need to handle commit, rollbacks the same as the other state machine state, to keep it consistent. So better to be handled in existing store directly.
but we can cache the whole thing in memory on start up, so duplication checking is fast.

Reply:
we do something like this in store, but not sure its exposed to things out side of store currently. The main worry i have is around upgrades, where everyone stops and starts, then the map would be empty. Looking back at the x blocks would be best but since we dont know if everyone has it, we could end up in a bad situation

Reply:
yeah, we need some support from storage layer.